File: RateLimitterJava/src/service/RateLimitService.java
package service;

import factory.RateLimitServiceBuilder;
import request.RequestContext;
import strategy.RateLimitStrategy;

public class RateLimitService {
    private final RateLimitServiceBuilder rateLimitServiceBuilder;

    public RateLimitService(RateLimitServiceBuilder rateLimitServiceBuilder) {
        this.rateLimitServiceBuilder = rateLimitServiceBuilder;
    }

    public boolean isRequestAllowed(String strategyName, RequestContext requestContext, int userId) {
        RateLimitStrategy strategy = rateLimitServiceBuilder.getRateLimitStrategyOrchestrator(userId, "/fund-transfer");
        return strategy.isRequestAllowed(requestContext);
    }
}

File: RateLimitter/service/RateLimitService.kt
package service

import entity.TokenBucket
import request.RequestContext
import strategy.BucketTokenStrategy
import strategy.SlidingWindowStrategy
import java.time.Instant
import java.util.LinkedList
import java.util.Queue

class RateLimitService(
    private val slidingWindowStrategy: SlidingWindowStrategy,
    private val tokenBucketStrategy: BucketTokenStrategy
) {
    val timeWindow = 1
    val maxAllowedRequest = 10

    val userRequestLog = mutableMapOf<String, Queue<Long>>()

    fun isRequestAllowed(typeOfRateLimit: Int, userId: String): Boolean {
        val rateLimiterStrategy = if(typeOfRateLimit == 1){
            tokenBucketStrategy
        } else {
            slidingWindowStrategy
        }

        val requestBody = RequestContext(
            userId =  userId,
            timeStamp = Instant.now().toEpochMilli()
        )
       return rateLimiterStrategy.isRequestAllowed(requestBody)
    }
}